File: apps/Server/src/core/services/transaction_service.py
"""Transaction service for business logic."""

from typing import List, Optional, Tuple
from uuid import UUID

from sqlalchemy.orm import Session

from src.interface.transaction_dto import (
    TransactionCreateDTO,
    TransactionFilterDTO,
    TransactionUpdateDTO,
)
from src.models.transaction import Transaction
from src.repository.transaction_repository import transaction_repository


class TransactionService:
    """Service for transaction business logic."""

    def create_transaction(
        self,
        db: Session,
        user_id: UUID,
        data: TransactionCreateDTO,
    ) -> Transaction:
        """
        Create a new transaction.

        Args:
            db: Database session
            user_id: ID of the user creating the transaction
            data: Transaction creation data

        Returns:
            Created Transaction object
        """
        print(f"INFO [TransactionService]: Creating transaction for entity {data.entity_id}")
        transaction = transaction_repository.create_transaction(
            db=db,
            entity_id=data.entity_id,
            category_id=data.category_id,
            user_id=user_id,
            amount=data.amount,
            type=data.type,
            description=data.description,
            transaction_date=data.date,
            notes=data.notes,
        )
        print(f"INFO [TransactionService]: Transaction {transaction.id} created successfully")
        return transaction

    def get_transaction(
        self,
        db: Session,
        transaction_id: UUID,
        entity_id: UUID,
    ) -> Optional[Transaction]:
        """
        Get a transaction by ID, validating entity ownership.

        Args:
            db: Database session
            transaction_id: Transaction UUID
            entity_id: Entity UUID for validation

        Returns:
            Transaction object if found and belongs to entity, None otherwise
        """
        print(f"INFO [TransactionService]: Getting transaction {transaction_id}")
        transaction = transaction_repository.get_transaction_by_id(db, transaction_id)
        if transaction and transaction.entity_id != entity_id:
            print(f"ERROR [TransactionService]: Transaction {transaction_id} does not belong to entity {entity_id}")
            return None
        return transaction

    def list_transactions(
        self,
        db: Session,
        entity_id: UUID,
        filters: Optional[TransactionFilterDTO] = None,
        skip: int = 0,
        limit: int = 100,
    ) -> Tuple[List[Transaction], int]:
        """
        List transactions for an entity with pagination.

        Args:
            db: Database session
            entity_id: Entity UUID to filter by
            filters: Optional filter criteria
            skip: Number of records to skip
            limit: Maximum number of records to return

        Returns:
            Tuple of (list of transactions, total count)
        """
        print(f"INFO [TransactionService]: Listing transactions for entity {entity_id}")
        transactions = transaction_repository.get_transactions_by_entity(
            db=db,
            entity_id=entity_id,
            filters=filters,
            skip=skip,
            limit=limit,
        )
        total = transaction_repository.count_transactions_by_entity(
            db=db,
            entity_id=entity_id,
            filters=filters,
        )
        print(f"INFO [TransactionService]: Found {len(transactions)} transactions (total: {total})")
        return transactions, total

    def update_transaction(
        self,
        db: Session,
        transaction_id: UUID,
        entity_id: UUID,
        data: TransactionUpdateDTO,
    ) -> Optional[Transaction]:
        """
        Update an existing transaction.

        Args:
            db: Database session
            transaction_id: Transaction UUID
            entity_id: Entity UUID for validation
            data: Update data

        Returns:
            Updated Transaction object if found and updated, None otherwise
        """
        print(f"INFO [TransactionService]: Updating transaction {transaction_id}")
        transaction = transaction_repository.get_transaction_by_id(db, transaction_id)
        if not transaction:
            print(f"ERROR [TransactionService]: Transaction {transaction_id} not found")
            return None
        if transaction.entity_id != entity_id:
            print(f"ERROR [TransactionService]: Transaction {transaction_id} does not belong to entity {entity_id}")
            return None

        # Update fields if provided
        if data.category_id is not None:
            transaction.category_id = data.category_id
        if data.amount is not None:
            transaction.amount = data.amount
        if data.type is not None:
            transaction.type = data.type
        if data.description is not None:
            transaction.description = data.description
        if data.date is not None:
            transaction.date = data.date
        if data.notes is not None:
            transaction.notes = data.notes

        updated = transaction_repository.update_transaction(db, transaction)
        print(f"INFO [TransactionService]: Transaction {transaction_id} updated successfully")
        return updated

    def delete_transaction(
        self,
        db: Session,
        transaction_id: UUID,
        entity_id: UUID,
    ) -> bool:
        """
        Delete a transaction.

        Args:
            db: Database session
            transaction_id: Transaction UUID
            entity_id: Entity UUID for validation

        Returns:
            True if deleted, False if not found or not owned
        """
        print(f"INFO [TransactionService]: Deleting transaction {transaction_id}")
        transaction = transaction_repository.get_transaction_by_id(db, transaction_id)
        if not transaction:
            print(f"ERROR [TransactionService]: Transaction {transaction_id} not found")
            return False
        if transaction.entity_id != entity_id:
            print(f"ERROR [TransactionService]: Transaction {transaction_id} does not belong to entity {entity_id}")
            return False

        transaction_repository.delete_transaction(db, transaction)
        print(f"INFO [TransactionService]: Transaction {transaction_id} deleted successfully")
        return True


# Singleton instance
transaction_service = TransactionService()

File: apps/Client/src/services/transactionService.ts
import { apiClient } from '@/api/clients'
import type {
  Transaction,
  TransactionCreate,
  TransactionUpdate,
  TransactionFilters,
  TransactionListResponse,
} from '@/types'

/**
 * Transaction service for CRUD operations on transactions.
 */
export const transactionService = {
  /**
   * Create a new transaction.
   * @param data - Transaction creation data
   * @returns Created transaction
   */
  async create(data: TransactionCreate): Promise<Transaction> {
    console.log('INFO [TransactionService]: Creating transaction for entity:', data.entity_id)
    try {
      const response = await apiClient.post<Transaction>('/transactions/', data)
      console.log('INFO [TransactionService]: Transaction created:', response.data.id)
      return response.data
    } catch (error) {
      console.error('ERROR [TransactionService]: Failed to create transaction:', error)
      throw error
    }
  },

  /**
   * List transactions for an entity with optional filters.
   * @param entityId - Entity ID to filter by
   * @param filters - Optional filter criteria
   * @param skip - Number of records to skip
   * @param limit - Maximum records to return
   * @returns List of transactions with total count
   */
  async list(
    entityId: string,
    filters?: TransactionFilters,
    skip = 0,
    limit = 100
  ): Promise<TransactionListResponse> {
    console.log('INFO [TransactionService]: Listing transactions for entity:', entityId)
    try {
      const params = new URLSearchParams()
      params.append('entity_id', entityId)
      params.append('skip', skip.toString())
      params.append('limit', limit.toString())

      if (filters?.start_date) {
        params.append('start_date', filters.start_date)
      }
      if (filters?.end_date) {
        params.append('end_date', filters.end_date)
      }
      if (filters?.category_id) {
        params.append('category_id', filters.category_id)
      }
      if (filters?.type) {
        params.append('type', filters.type)
      }

      const response = await apiClient.get<TransactionListResponse>(
        `/transactions/?${params.toString()}`
      )
      console.log(
        'INFO [TransactionService]: Fetched',
        response.data.transactions.length,
        'transactions (total:',
        response.data.total,
        ')'
      )
      return response.data
    } catch (error) {
      console.error('ERROR [TransactionService]: Failed to list transactions:', error)
      throw error
    }
  },

  /**
   * Get a single transaction by ID.
   * @param transactionId - Transaction ID
   * @param entityId - Entity ID for validation
   * @returns Transaction data
   */
  async get(transactionId: string, entityId: string): Promise<Transaction> {
    console.log('INFO [TransactionService]: Getting transaction:', transactionId)
    try {
      const response = await apiClient.get<Transaction>(
        `/transactions/${transactionId}?entity_id=${entityId}`
      )
      console.log('INFO [TransactionService]: Transaction fetched:', response.data.id)
      return response.data
    } catch (error) {
      console.error('ERROR [TransactionService]: Failed to get transaction:', error)
      throw error
    }
  },

  /**
   * Update an existing transaction.
   * @param transactionId - Transaction ID to update
   * @param entityId - Entity ID for validation
   * @param data - Update data
   * @returns Updated transaction
   */
  async update(
    transactionId: string,
    entityId: string,
    data: TransactionUpdate
  ): Promise<Transaction> {
    console.log('INFO [TransactionService]: Updating transaction:', transactionId)
    try {
      const response = await apiClient.put<Transaction>(
        `/transactions/${transactionId}?entity_id=${entityId}`,
        data
      )
      console.log('INFO [TransactionService]: Transaction updated:', response.data.id)
      return response.data
    } catch (error) {
      console.error('ERROR [TransactionService]: Failed to update transaction:', error)
      throw error
    }
  },

  /**
   * Delete a transaction.
   * @param transactionId - Transaction ID to delete
   * @param entityId - Entity ID for validation
   */
  async delete(transactionId: string, entityId: string): Promise<void> {
    console.log('INFO [TransactionService]: Deleting transaction:', transactionId)
    try {
      await apiClient.delete(`/transactions/${transactionId}?entity_id=${entityId}`)
      console.log('INFO [TransactionService]: Transaction deleted:', transactionId)
    } catch (error) {
      console.error('ERROR [TransactionService]: Failed to delete transaction:', error)
      throw error
    }
  },
}

export default transactionService
